compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/model/BIRNode.java

```diff
@@ -817,6 +869,57 @@ protected BIRDocumentableNode(Location pos) {
         public void setMarkdownDocAttachment(MarkdownDocAttachment markdownDocAttachment) {
             this.markdownDocAttachment = markdownDocAttachment;
         }
+
+        public void addChildNode(BIRDocumentableNode childNode) {
+            if (childNode == null) {
+                return;
+            }
+            childNodes.add(childNode);
+
+            addParent(childNode, this);
+
+            if (this.usedState == UsedState.USED) {
+                // It is possible to omit unused modules from codegen if they are identified in the analyzer phase
+                childNode.markSelfAndChildrenAsUsed();
+            }
+        }
+
+        private void addParent(BIRDocumentableNode childNode, BIRDocumentableNode parentNode) {
+            childNode.parentNodes.add(parentNode);
+        }
+
+        public void markSelfAndChildrenAsUsed() {
+            if (usedState != UsedState.USED) {
+                usedState = UsedState.USED;
+                for (BIRDocumentableNode childNode : this.childNodes) {
+                    if (childNode != null) {
+                        childNode.markSelfAndChildrenAsUsed();
+                    }
+                }
+            }
```

Comment:
IMO it is cleaner to early return and avoid unnecessary indentation here
```suggestion
            if (usedState == UsedState.USED) {
                return;
            }
            usedState = UsedState.USED;
            for (BIRDocumentableNode childNode : this.childNodes) {
                if (childNode != null) {
                      childNode.markSelfAndChildrenAsUsed();
                 }
            }
```